src/dhcp/clientv4.rs

```diff
@@ -0,0 +1,458 @@
+
+        Client {
+            state: ClientState::Discovering,
+            raw_handle,
+            next_egress: now,
+            transaction_id: 1,
+            dns_servers: [None; 3],
+        }
+    }
+
+    /// When to send next packet
+    ///
+    /// Useful for suspending execution after polling.
+    pub fn next_poll(&self, now: Instant) -> Duration {
+        self.next_egress - now
+    }
+
+    /// Process incoming packets on the contained RawSocket, and send
+    /// DHCP requests when timeouts are ready.
+    pub fn poll<DeviceT: for<'d> Device<'d>>(&mut self, iface: &mut Interface<DeviceT>, sockets: &mut SocketSet, now: Instant) -> Result<()> {
+        let checksum_caps = ChecksumCapabilities::default();
+        let mut raw_socket = sockets.get::<RawSocket>(self.raw_handle);
+
+        // Process incoming
+        {
+            match raw_socket.recv()
+                .and_then(|packet| parse_udp(packet, &checksum_caps)) {
+                    Ok((src, dst, payload)) =>
+                        self.ingress(iface, now, payload, &src, &dst),
+                    Err(Error::Exhausted) =>
+                        (),
+                    Err(e) =>
+                        return Err(e),
+                }
+        }
+
+        // Send requests
+        if raw_socket.can_send() && now >= self.next_egress {
+            self.egress(iface, &mut *raw_socket, &checksum_caps, now)?;
+        }
+
+        Ok(())
+    }
+
+    fn ingress<DeviceT: for<'d> Device<'d>>(&mut self, iface: &mut Interface<DeviceT>, now: Instant, data: &[u8], src: &IpEndpoint, dst: &IpEndpoint) {
+        if src.port != UDP_SERVER_PORT ||
+           dst.port != UDP_CLIENT_PORT { return }
+        let src_ip = match src.addr {
+            IpAddress::Ipv4(src_addr) => src_addr,
+            _ => return,
+        };
+
+        let dhcp_packet = match DhcpPacket::new_checked(data) {
+            Ok(dhcp_packet) => dhcp_packet,
+            Err(_) => return,
+        };
+        let dhcp_repr = match DhcpRepr::parse(&dhcp_packet) {
+            Ok(dhcp_repr) => dhcp_repr,
+            Err(_) => return,
+        };
+        let mac = iface.ethernet_addr();
+        if dhcp_repr.client_hardware_address != mac { return }
+        if dhcp_repr.transaction_id != self.transaction_id { return }
+        let server_identifier = match dhcp_repr.server_identifier {
+            Some(server_identifier) => server_identifier,
+            None => return,
+        };
+        net_debug!("DHCP recv {:?} from {} ({})", dhcp_repr.message_type, src, server_identifier);
+
+        if (dhcp_repr.message_type == DhcpMessageType::Offer ||
+            dhcp_repr.message_type == DhcpMessageType::Ack) &&
+           dhcp_repr.your_ip != Ipv4Address::UNSPECIFIED {
+               let prefix_len = dhcp_repr.subnet_mask
+                   .and_then(|mask| IpAddress::Ipv4(mask).to_prefix_len());
+               // Replace the first IP address
+               iface.update_ip_addrs(|addrs| {
```

Comment:
Please make this a callback. Hardcoding replacement of the first IP address is not very good, but silently doing nothing if a suitable address isn't found is much worse.

Reply:
As a callback interface is public API, there are a few design aspects to keep in mind:
- Corresponding configuration options may exist for IPv6 one day
- As they are used for ethernet interface configuration, that should ideally accept those configuration options

Your guidance is welcome.

Reply:
DHCP and DHCPv6 are in effect completely different protocols, aren't they? And often you'd have a combination of DHCP and SLAAC, not DHCPv6.

I think it's fine if the callback interface is DHCP(v4)-specific. Even if that turns out to be wrong, it is a relatively small part of the API and we can probably just break it.